### PR TITLE
index processing guidelines

### LINKED ISSUE #1185: Write guide to index file processing
/assign

### PR DESCRIPTION
**Pull request by mikehelmick**
Fixes #1185 

## Proposed Changes

* add guide for index.txt processing

**Release Note**

```release-note
Guide for app developers in processing the export index file.
```

/hold
Want LGTM from @peetsasaki and @flagxor (both) before submit.

### CODE REVIEW

File: docs/getting-started/downloading-export-batches-keys.md [hunk 1]
```diff
@@ -11,6 +11,84 @@ server. Keys can be grouped by region and the frequency of batching is
...
 
+# Consuming Export Batches
+
+## Index File Structure
+
+The index.txt file indicates which export files are available for a mobile device
+to consume. The index file contains relative path names for available archives.
+A client is typically given two configuration items for pointing at a server
+
+* The absolute location of the index.txt file
```
Comment (sethvargo) [inline docs/getting-started/downloading-export-batches-keys.md:22]: Should we make a note that app developers are encouraged to use runtime configuration and **NOT** hardcode these values?

File: docs/getting-started/downloading-export-batches-keys.md [hunk 2]
```diff
@@ -11,6 +11,84 @@ server. Keys can be grouped by region and the frequency of batching is
...
+
+And these are being arrived at by appending the contents of the index file (one per line)
+to the base path provided by the server operator.
+
+### Index Processing Guideliens
+
+* Do not make any assumptions or dependencies on the layout of the .zip filenames themselves.
+* Do not make any assumptions about the depth of the ralative path in the index file. 
+* The index.txt will always be ordered from oldest file to newest file.
+* Each file only needs to be processed once.
```
Comment (sethvargo) [inline docs/getting-started/downloading-export-batches-keys.md:49]: \* Do not assume files will end with `.zip`, but _do_ assume that all files are in zip format.

File: docs/getting-started/downloading-export-batches-keys.md [hunk 3]
```diff
@@ -11,6 +11,84 @@ server. Keys can be grouped by region and the frequency of batching is
...
+1. `https://cdn.example.com/24h/region/file1.zip`
+1. `https://cdn.example.com/24h/region/file21.zip`
+1. `https://cdn.example.com/24h/region/other_export_.zip`
+
+And these are being arrived at by appending the contents of the index file (one per line)
+to the base path provided by the server operator.
+
+### Index Processing Guideliens
+
+* Do not make any assumptions or dependencies on the layout of the .zip filenames themselves.
```
Comment (sethvargo) [inline docs/getting-started/downloading-export-batches-keys.md:46]: Let's add something about the character set/encoding? US-ASCII, UTF-8

File: docs/getting-started/downloading-export-batches-keys.md [hunk 4]
```diff
@@ -11,6 +11,84 @@ server. Keys can be grouped by region and the frequency of batching is
...
+
+There are two modes of index file processing that meet these guidelines.
+
+#### Last seen method (most common):
+
+The most common method (in use by the Android reference application) is for the client
+to remember the last filename that it has seen in an index.txt file.
+
+On next processing, if that line is seen in the file, the application will process
+from the next file to the end of the index. If that lien is not present in the file,
```
Comment (sethvargo) [inline docs/getting-started/downloading-export-batches-keys.md:59]:
```suggestion
from the next file to the end of the index. If that line is not present in the file,
```

**Comment by sethvargo** [comment]
/retest

@mikehelmick made a few small nits and there's one typo

File: docs/getting-started/downloading-export-batches-keys.md [hunk 5]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+Then the following 3 archives are are being offered to clients:
+
+1. `https://cdn.example.com/24h/region/file1.zip`
+1. `https://cdn.example.com/24h/region/file21.zip`
+1. `https://cdn.example.com/24h/region/other_export_.zip`
+
+And these are being arrived at by appending the contents of the index file (one per line)
+to the base path provided by the server operator.
+
+### Index Processing Guideliens
```
Comment (whaught) [inline docs/getting-started/downloading-export-batches-keys.md:48]:
```suggestion
### Index Processing Guidelines
```

File: docs/getting-started/downloading-export-batches-keys.md [hunk 6]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
 
+# Consuming Export Batches
+
+## Index File Structure
+
+The index.txt file indicates which export files are available for a mobile device
+to consume. The index file contains relative path names for available archives.
+A client is typically given two configuration items for pointing at a server.
+
+Mobile app developers are strongly encouraged to make these two values remotly
```
Comment (whaught) [inline docs/getting-started/downloading-export-batches-keys.md:22]:
```suggestion
Mobile app developers are strongly encouraged to make these two values remotely
```

File: docs/getting-started/downloading-export-batches-keys.md [hunk 7]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+On next processing, if that line is seen in the file, the application will process
+from the next file to the end of the index. If that line is not present in the file,
+then the application will restart processing of the index from the beginning (line 0).
+
+When done, the application always remembers the last filename processed.
+
+#### Set method:
+
+The client application should treate absolute URLs of the exports indicated in the the
+index.txt file contents as a set. The client should store the previsuly known state in
```
Comment (whaught) [inline docs/getting-started/downloading-export-batches-keys.md:74]:
```suggestion
index.txt file contents as a set. The client should store the previously known state in
```

File: docs/getting-started/downloading-export-batches-keys.md [hunk 8]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+to the operating system.
+
+Practically speaking: currently all export files produced by this system will have
+`batchNum` of `1` and `batchSize` of `1`. In the event a time period is too 
+large to find in one file according to the max records setting, there will
+be multiple files with the same `startTimestamp` but different `endTimestamps`. This
+is done to ensure successful processing based on how both iOS and Android
+handle export processing.
+
+# Configurating Export Batches
```
Comment (whaught) [inline docs/getting-started/downloading-export-batches-keys.md:97]:
```suggestion
# Configuring Export Batches
```

File: docs/getting-started/downloading-export-batches-keys.md [hunk 9]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+region/other_export_.zip
+```
+
+Then the following 3 archives are are being offered to clients:
+
+1. `https://cdn.example.com/24h/region/file1.zip`
+1. `https://cdn.example.com/24h/region/file21.zip`
+1. `https://cdn.example.com/24h/region/other_export_.zip`
+
+And these are being arrived at by appending the contents of the index file (one per line)
```
Comment (crwilcox) [inline docs/getting-started/downloading-export-batches-keys.md:45]:
```suggestion
And these are derived by appending the contents of the index file (one per line)
```
Or constructed? "being arrived at" is a bit hard to read.

File: docs/getting-started/downloading-export-batches-keys.md [hunk 10]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+When done, the application always remembers the last filename processed.
+
+#### Set method:
+
+The client application should treate absolute URLs of the exports indicated in the the
+index.txt file contents as a set. The client should store the previsuly known state in
+local storage.
+
+If differences are seen, the clients should download the items missing from the set in top
+to bottom order (oldest to newest) and hand them to the operating system for matching.
```
Comment (crwilcox) [inline docs/getting-started/downloading-export-batches-keys.md:78]:
```suggestion
to bottom order (oldest to newest) and provide them to the operating system for matching.
```
or pass, which is used later as the action

File: docs/getting-started/downloading-export-batches-keys.md [hunk 11]
```diff
@@ -11,6 +11,91 @@ server. Keys can be grouped by region and the frequency of batching is
...
+
+The new set of absolute URLs should then be retained for processing next time around.
+
+## Export File Concepts
+
+As the protocol has changed over the deployment of various exposure notifications
+installations, we have made a decisions to drop the usage of a batch
+being split across multiple files, and show all batches as single file size only.
+This reduces the possilibty that a client app will pass un-parsable batch files
```
Comment (crwilcox) [inline docs/getting-started/downloading-export-batches-keys.md:87]:
```suggestion
This reduces the possibility that a client app will pass un-parsable batch files
```

**Comment by mikehelmick** [comment]
updated

**Pull request merged**